File: 1024_Video_Stitching.py
class Solution:
    def videoStitching(self, clips: List[List[int]], T: int) -> int:
        dp = [T + 1] * (T + 1)
        dp[0] = 0
        for i in range(T + 1):
            for c in clips:
                if c[0] <= i <= c[1]:
                    dp[i] = min(dp[i], dp[c[0]] + 1)
        return -1 if dp[T] == T + 1 else dp[T]

class Solution:
    def videoStitching(self, clips: List[List[int]], T: int) -> int:
        start, end, res = -1, 0, 0
        for s, e in sorted(clips):
            if end >= T or s > end:
                break
            elif start < s <= end:
                res, start = res + 1, end
            end = max(end, e)
        return res if end >= T else -1

File: 1024_Video_Stitching.cpp
class Solution {
public:
    int videoStitching(vector<vector<int>>& clips, int T) {
        vector<int> max_range (T + 1, 0);
        for (auto clip : clips) {
            if (clip[0] < T + 1) {
                max_range[clip[0]] = max(max_range[clip[0]],
                                         min(clip[1] - clip[0], T - clip[0]));
            }
        }
        int start = 0, end = 0, step = 0;
        while (end < T) {
            ++step;
            int tmp = end;
            for (int i = start; i <= tmp; ++i) {
                end = max(end, i + max_range[i]);
            }
            start = tmp;
            if (start == end) return -1;
        }
        return step;
    }
};
